File: infrastructure/solvers/fdm_heat_solver_3d.py
"""
3D Finite Difference Method Heat Solver.

Implements IHeatSolver for full 3D transient heat conduction:
    ∂T/∂t = κ (∂²T/∂x² + ∂²T/∂y² + ∂²T/∂z²)

Numerical scheme: Forward-Time Central-Space (FTCS) explicit method.

CIA Triad:
- Integrity: Automatic CFL stability check before solving.
- Availability: Grid size limit to prevent memory exhaustion (DoS).

DRY: Beam propagation logic imported from a shared helper,
     not duplicated from the laser module.
"""

import math
import time as time_module

import numpy as np

from domain.entities import (
    Material,
    LaserBeam,
    SimulationResult,
    Trajectory,
)
from domain.exceptions import GridTooLargeError, SolverInstabilityError
from domain.ports import IHeatSolver, ILaserSource


# ============================================================================
# Shared Helper: Beam Propagation (DRY — used by solver only)
# ============================================================================

def compute_spot_radius_at_z(z: float, laser: LaserBeam) -> float:
    """
    Gaussian beam propagation: w(z) = w0 · √(1 + (z/z_R)²).

    Extracted as a pure function to avoid duplication (DRY).
    """
    z_r = laser.rayleigh_range
    return laser.spot_radius * math.sqrt(1 + (z / z_r) ** 2)


# ============================================================================
# Constants
# ============================================================================

MAX_GRID_CELLS = 5_000_000  # Safety limit (Availability — DoS prevention)


# ============================================================================
# 3D FDM Solver
# ============================================================================

class FDMHeatSolver3D(IHeatSolver):
    """
    3D transient heat solver using the explicit FTCS finite difference method.

    Solves: ∂T/∂t = κ · ∇²T  with surface heat flux from laser.

    Boundary Conditions:
      - Top surface (z=0): Neumann BC (laser heat flux during scan)
      - Bottom (z=max): Adiabatic (∂T/∂z = 0) or Dirichlet (T_ambient)
      - Lateral boundaries: Adiabatic (∂T/∂x = 0, ∂T/∂y = 0)
    """

    def __init__(
        self,
        laser_source: ILaserSource,
        bottom_bc: str = "adiabatic",
    ):
        """
        Args:
            laser_source: Injected beam profile strategy (DIP).
            bottom_bc: Bottom boundary condition ('adiabatic' or 'fixed').
        """
        self._laser_source = laser_source
        self._bottom_bc = bottom_bc

    @property
    def name(self) -> str:
        return "FDM-3D-FTCS"

    def solve(
        self,
        material: Material,
        laser: LaserBeam,
        trajectory: Trajectory,
        grid_size: tuple = (0.010, 0.010, 0.002),
        resolution: float = 100e-6,
    ) -> SimulationResult:
        """
        Run the full 3D heat simulation.

        Args:
            material: Target material properties.
            laser: Laser beam configuration.
            trajectory: Motion path (time, x, y, z arrays).
            grid_size: (Lx, Ly, Lz) domain size in meters.
                       Default: 10mm × 10mm × 2mm.
            resolution: Spatial step Δx = Δy = Δz [m].
                        Default: 100 µm.

        Returns:
            SimulationResult with 3D temperature snapshot and metrics.
        """
        start_time = time_module.perf_counter()

        # --- Grid Setup ---
        lx, ly, lz = grid_size
        dx = resolution
        dy = resolution
        dz = resolution

        nx = max(int(lx / dx), 3)
        ny = max(int(ly / dy), 3)
        nz = max(int(lz / dz), 3)

        total_cells = nx * ny * nz
        self._check_grid_size(total_cells)

        kappa = material.thermal_diffusivity

        # --- CFL Stability Check ---
        dt_max = dx ** 2 / (6.0 * kappa)  # 3D stability limit
        dt_trajectory = self._compute_trajectory_dt(trajectory)
        dt = min(dt_max * 0.8, dt_trajectory)  # 80% safety margin

        if dt <= 0:
            raise SolverInstabilityError(
                f"Invalid time step dt={dt:.2e}s. Check material diffusivity."
            )

        # --- Initialize Temperature Field ---
        t_field = np.full((nx, ny, nz), material.t_ambient, dtype=np.float64)

        # Coordinate arrays
        x_coords = np.linspace(0, lx, nx)
        y_coords = np.linspace(0, ly, ny)
        z_coords = np.linspace(0, lz, nz)

        # 2D meshgrid for surface (z=0)
        x_mesh, y_mesh = np.meshgrid(x_coords, y_coords, indexing="ij")

        # --- Cumulative Fluence Map ---
        fluence_map = np.zeros((nx, ny), dtype=np.float64)

        # --- Precompute Coefficients ---
        rx = kappa * dt / dx ** 2
        ry = kappa * dt / dy ** 2
        rz = kappa * dt / dz ** 2

        # --- Time-Stepping Loop ---
        n_steps = trajectory.n_points
        traj_time = np.array(trajectory.time)
        traj_x = np.array(trajectory.x)
        traj_y = np.array(trajectory.y)
        traj_z = np.array(trajectory.z)

        for step_idx in range(n_steps - 1):
            # Determine sub-steps for this trajectory segment
            seg_dt = traj_time[step_idx + 1] - traj_time[step_idx]
            if seg_dt <= 0:
                continue

            n_sub = max(int(np.ceil(seg_dt / dt)), 1)
            actual_dt = seg_dt / n_sub

            # Recalculate coefficients for actual_dt
            rx_a = kappa * actual_dt / dx ** 2
            ry_a = kappa * actual_dt / dy ** 2
            rz_a = kappa * actual_dt / dz ** 2

            # Beam position (interpolate within segment)
            beam_x = traj_x[step_idx]
            beam_y = traj_y[step_idx]
            beam_z_distance = traj_z[step_idx]

            # Compute spot radius at this defocus distance
            w_z = compute_spot_radius_at_z(beam_z_distance, laser)

            # Compute surface intensity distribution
            intensity_2d = self._laser_source.compute_intensity(
                x_mesh, y_mesh, beam_x, beam_y, w_z, laser.power
            )

            # Apply material absorption
            absorbed_intensity = material.absorption * intensity_2d

            # Accumulate fluence
            fluence_map += absorbed_intensity * seg_dt

            # Surface heat flux: q = α · I(x,y) [W/m²]
            # Neumann BC at z=0: dT/dz|_{z=0} = -q/k
            surface_flux_temperature_rate = (
                absorbed_intensity / (material.density * material.specific_heat * dz)
            )

            for _ in range(n_sub):
                t_field = self._diffusion_step(
                    t_field, rx_a, ry_a, rz_a,
                    surface_flux_temperature_rate, actual_dt,
                    material.t_ambient,
                )

        # --- Extract Results ---
        elapsed = time_module.perf_counter() - start_time

        peak_temp = float(np.max(t_field))
        peak_fluence = float(np.max(fluence_map))

        melt_depth = self._find_threshold_depth(
            t_field, z_coords, material.t_melt
        )
        vap_depth = self._find_threshold_depth(
            t_field, z_coords, material.t_vaporization
        )

        total_energy = float(np.sum(fluence_map) * dx * dy)

        return SimulationResult(
            temperature_field=t_field.tolist(),
            fluence_map=fluence_map.tolist(),
            x_coords=x_coords.tolist(),
            y_coords=y_coords.tolist(),
            z_coords=z_coords.tolist(),
            time_coords=list(trajectory.time),
            peak_temperature_celsius=peak_temp,
            peak_fluence_j_per_m2=peak_fluence,
            melt_depth_m=melt_depth,
            vaporization_depth_m=vap_depth,
            total_energy_j=total_energy,
            material_name=material.name,
            solver_name=self.name,
            duration_seconds=elapsed,
        )

    # ===== Private Methods (SRP: each handles one sub-task) =====

    def _diffusion_step(
        self,
        t_field: np.ndarray,
        rx: float,
        ry: float,
        rz: float,
        surface_source: np.ndarray,
        dt: float,
        t_ambient: float,
    ) -> np.ndarray:
        """
        Single FTCS diffusion step in 3D.

        Uses numpy slicing for vectorized computation.
        """
        nx, ny, nz = t_field.shape
        t_new = t_field.copy()

        # Interior points: 3D Laplacian
        t_new[1:-1, 1:-1, 1:-1] = t_field[1:-1, 1:-1, 1:-1] + (
            rx * (t_field[2:, 1:-1, 1:-1] + t_field[:-2, 1:-1, 1:-1]
                  - 2 * t_field[1:-1, 1:-1, 1:-1])
            + ry * (t_field[1:-1, 2:, 1:-1] + t_field[1:-1, :-2, 1:-1]
                    - 2 * t_field[1:-1, 1:-1, 1:-1])
            + rz * (t_field[1:-1, 1:-1, 2:] + t_field[1:-1, 1:-1, :-2]
                    - 2 * t_field[1:-1, 1:-1, 1:-1])
        )

        # --- Boundary Conditions ---

        # Top surface (z=0): Apply laser heat flux
        t_new[:, :, 0] += surface_source * dt

        # Also apply diffusion at surface (z=0) using ghost node = z=1
        t_new[1:-1, 1:-1, 0] += rz * (
            t_field[1:-1, 1:-1, 1] - t_field[1:-1, 1:-1, 0]
        )

        # Bottom boundary (z = nz-1)
        if self._bottom_bc == "adiabatic":
            t_new[:, :, -1] = t_new[:, :, -2]  # Zero gradient
        else:
            t_new[:, :, -1] = t_ambient  # Fixed temperature

        # Lateral boundaries: Adiabatic (mirror)
        t_new[0, :, :] = t_new[1, :, :]
        t_new[-1, :, :] = t_new[-2, :, :]
        t_new[:, 0, :] = t_new[:, 1, :]
        t_new[:, -1, :] = t_new[:, -2, :]

        return t_new

    @staticmethod
    def _check_grid_size(total_cells: int) -> None:
        """DoS prevention: reject grids that would exhaust memory."""
        if total_cells > MAX_GRID_CELLS:
            raise GridTooLargeError(total_cells, MAX_GRID_CELLS)

    @staticmethod
    def _compute_trajectory_dt(trajectory: Trajectory) -> float:
        """Compute the smallest time step from the trajectory."""
        times = trajectory.time
        min_dt = float("inf")
        for i in range(len(times) - 1):
            seg_dt = times[i + 1] - times[i]
            if seg_dt > 0:
                min_dt = min(min_dt, seg_dt)
        return min_dt if min_dt != float("inf") else 1e-4

    @staticmethod
    def _find_threshold_depth(
        t_field: np.ndarray, z_coords: np.ndarray, threshold: float
    ):
        """Find maximum depth where temperature exceeds threshold."""
        # Find along z-axis (axis=2), check max temperature at each z
        max_temp_per_z = np.max(t_field, axis=(0, 1))

        # Find deepest z where temp > threshold
        exceeds = np.where(max_temp_per_z > threshold)[0]
        if len(exceeds) == 0:
            return None
        return float(z_coords[exceeds[-1]])

File: UnityHVOF/Assets/Scripts/Infrastructure/Solvers/FDMHeatSolver3D.cs
// ============================================================================
// 3D Finite Difference Method Heat Solver.
//
// Implements IHeatSolver for full 3D transient heat conduction:
//     ∂T/∂t = κ (∂²T/∂x² + ∂²T/∂y² + ∂²T/∂z²)
//
// Numerical scheme: Forward-Time Central-Space (FTCS) explicit method.
//
// CIA Triad:
//   - Integrity: Automatic CFL stability check before solving.
//   - Availability: Grid size limit to prevent memory exhaustion (DoS).
//
// DRY: Beam propagation logic is a static helper, not duplicated.
//
// Port of: infrastructure/solvers/fdm_heat_solver_3d.py (324 lines)
// ============================================================================

using System;
using System.Diagnostics;
using HVOFSim.Domain.Entities;
using HVOFSim.Domain.Exceptions;
using HVOFSim.Domain.Ports;

namespace HVOFSim.Infrastructure.Solvers
{
    /// <summary>
    /// 3D transient heat solver using the explicit FTCS finite difference method.
    /// Solves: ∂T/∂t = κ · ∇²T with surface heat flux from laser.
    /// </summary>
    public sealed class FDMHeatSolver3D : IHeatSolver
    {
        /// <summary>Safety limit (Availability — DoS prevention).</summary>
        private const int MaxGridCells = 5_000_000;

        private readonly ILaserSource _laserSource;
        private readonly string _bottomBc;

        /// <summary>
        /// Create a 3D FDM heat solver.
        /// </summary>
        /// <param name="laserSource">Injected beam profile strategy (DIP).</param>
        /// <param name="bottomBc">Bottom boundary condition: "adiabatic" or "fixed".</param>
        public FDMHeatSolver3D(ILaserSource laserSource, string bottomBc = "adiabatic")
        {
            _laserSource = laserSource;
            _bottomBc = bottomBc;
        }

        public string Name => "FDM-3D-FTCS";

        public SimulationResult Solve(
            Material material,
            LaserBeam laser,
            Trajectory trajectory,
            double[] gridSize = null,
            double resolution = 100e-6)
        {
            gridSize ??= new[] { 0.010, 0.010, 0.002 };
            var stopwatch = Stopwatch.StartNew();

            // --- Grid Setup ---
            double lx = gridSize[0], ly = gridSize[1], lz = gridSize[2];
            double dx = resolution, dy = resolution, dz = resolution;

            int nx = Math.Max((int)(lx / dx), 3);
            int ny = Math.Max((int)(ly / dy), 3);
            int nz = Math.Max((int)(lz / dz), 3);

            int totalCells = nx * ny * nz;
            CheckGridSize(totalCells);

            double kappa = material.ThermalDiffusivity;

            // --- CFL Stability Check ---
            double dtMax = dx * dx / (6.0 * kappa); // 3D stability limit
            double dtTrajectory = ComputeTrajectoryDt(trajectory);
            double dt = Math.Min(dtMax * 0.8, dtTrajectory); // 80% safety margin

            if (dt <= 0)
                throw new SolverInstabilityException(
                    $"Invalid time step dt={dt:E2}s. Check material diffusivity.");

            // --- Initialize Temperature Field ---
            var tField = new double[nx, ny, nz];
            for (int i = 0; i < nx; i++)
                for (int j = 0; j < ny; j++)
                    for (int k = 0; k < nz; k++)
                        tField[i, j, k] = material.TAmbient;

            // Coordinate arrays
            double[] xCoords = Linspace(0, lx, nx);
            double[] yCoords = Linspace(0, ly, ny);
            double[] zCoords = Linspace(0, lz, nz);

            // 2D meshgrid for surface (z=0)
            var xMesh = new double[nx, ny];
            var yMesh = new double[nx, ny];
            for (int i = 0; i < nx; i++)
                for (int j = 0; j < ny; j++)
                {
                    xMesh[i, j] = xCoords[i];
                    yMesh[i, j] = yCoords[j];
                }

            // Cumulative fluence map
            var fluenceMap = new double[nx, ny];

            // --- Time-Stepping Loop ---
            int nSteps = trajectory.NPoints;
            long totalSubsteps = 0;
            const long MaxTotalSubsteps = 200_000; // Overall budget to prevent runaway

            for (int stepIdx = 0; stepIdx < nSteps - 1; stepIdx++)
            {
                double segDt = trajectory.Time[stepIdx + 1] - trajectory.Time[stepIdx];
                if (segDt <= 0) continue;

                int nSub = Math.Max((int)Math.Ceiling(segDt / dt), 1);
                
                // Prevent freeze on extreme per-segment substep counts
                if (nSub > 5000)
                {
                    throw new SolverInstabilityException(
                        $"Simulation parameters require {nSub} substeps per segment " +
                        $"(limit: 5000). Increase resolution, spot size, or scan speed.");
                }

                totalSubsteps += nSub;
                if (totalSubsteps > MaxTotalSubsteps)
                {
                    throw new SolverInstabilityException(
                        $"Total substeps ({totalSubsteps}) exceed budget ({MaxTotalSubsteps}). " +
                        $"The scan area is too large for the current resolution. " +
                        $"Try a smaller scan area or increase spot size / speed.");
                }

                double actualDt = segDt / nSub;

                double rxA = kappa * actualDt / (dx * dx);
                double ryA = kappa * actualDt / (dy * dy);
                double rzA = kappa * actualDt / (dz * dz);

                double beamX = trajectory.X[stepIdx];
                double beamY = trajectory.Y[stepIdx];
                double beamZDistance = trajectory.Z[stepIdx];

                // Compute spot radius at this defocus distance
                double wZ = ComputeSpotRadiusAtZ(beamZDistance, laser);

                // Compute surface intensity distribution
                var intensity2D = _laserSource.ComputeIntensity(
                    xMesh, yMesh, beamX, beamY, wZ, laser.Power);

                // Apply material absorption
                var absorbedIntensity = new double[nx, ny];
                for (int i = 0; i < nx; i++)
                    for (int j = 0; j < ny; j++)
                    {
                        absorbedIntensity[i, j] = material.Absorption * intensity2D[i, j];
                        fluenceMap[i, j] += absorbedIntensity[i, j] * segDt;
                    }

                // Surface heat flux rate
                var surfaceFluxRate = new double[nx, ny];
                double rhoTimeCpTimesDz = material.Density * material.SpecificHeat * dz;
                for (int i = 0; i < nx; i++)
                    for (int j = 0; j < ny; j++)
                        surfaceFluxRate[i, j] = absorbedIntensity[i, j] / rhoTimeCpTimesDz;

                for (int sub = 0; sub < nSub; sub++)
                {
                    tField = DiffusionStep(
                        tField, nx, ny, nz,
                        rxA, ryA, rzA,
                        surfaceFluxRate, actualDt,
                        material.TAmbient);
                }
            }

            // --- Extract Results ---
            stopwatch.Stop();
            double elapsed = stopwatch.Elapsed.TotalSeconds;

            double peakTemp = double.MinValue;
            double peakFluence = double.MinValue;

            for (int i = 0; i < nx; i++)
                for (int j = 0; j < ny; j++)
                {
                    if (fluenceMap[i, j] > peakFluence) peakFluence = fluenceMap[i, j];
                    for (int k = 0; k < nz; k++)
                        if (tField[i, j, k] > peakTemp) peakTemp = tField[i, j, k];
                }

            double? meltDepth = FindThresholdDepth(tField, nx, ny, nz, zCoords, material.TMelt);
            double? vapDepth = FindThresholdDepth(tField, nx, ny, nz, zCoords, material.TVaporization);

            double totalEnergy = 0;
            for (int i = 0; i < nx; i++)
                for (int j = 0; j < ny; j++)
                    totalEnergy += fluenceMap[i, j];
            totalEnergy *= dx * dy;

            return new SimulationResult(
                temperatureField: tField,
                fluenceMap: fluenceMap,
                xCoords: xCoords,
                yCoords: yCoords,
                zCoords: zCoords,
                timeCoords: trajectory.Time,
                peakTemperatureCelsius: peakTemp,
                peakFluenceJPerM2: peakFluence,
                meltDepthM: meltDepth,
                vaporizationDepthM: vapDepth,
                totalEnergyJ: totalEnergy,
                materialName: material.Name,
                solverName: Name,
                durationSeconds: elapsed);
        }

        // ===== Private Methods (SRP: each handles one sub-task) =====

        private double[,,] DiffusionStep(
            double[,,] tField, int nx, int ny, int nz,
            double rx, double ry, double rz,
            double[,] surfaceSource, double dt, double tAmbient)
        {
            var tNew = (double[,,])tField.Clone();

            // Interior points: 3D Laplacian
            for (int i = 1; i < nx - 1; i++)
                for (int j = 1; j < ny - 1; j++)
                    for (int k = 1; k < nz - 1; k++)
                    {
                        tNew[i, j, k] = tField[i, j, k]
                            + rx * (tField[i + 1, j, k] + tField[i - 1, j, k] - 2 * tField[i, j, k])
                            + ry * (tField[i, j + 1, k] + tField[i, j - 1, k] - 2 * tField[i, j, k])
                            + rz * (tField[i, j, k + 1] + tField[i, j, k - 1] - 2 * tField[i, j, k]);
                    }

            // Top surface (z=0): Apply laser heat flux
            for (int i = 0; i < nx; i++)
                for (int j = 0; j < ny; j++)
                    tNew[i, j, 0] += surfaceSource[i, j] * dt;

            // Diffusion at surface using ghost node = z=1
            for (int i = 1; i < nx - 1; i++)
                for (int j = 1; j < ny - 1; j++)
                    tNew[i, j, 0] += rz * (tField[i, j, 1] - tField[i, j, 0]);

            // Lateral (x-y) diffusion at the surface — the interior loop
            // starts at k=1 so the surface was missing this contribution,
            // causing heat to pile up at the irradiated cell instead of
            // spreading laterally which under-predicts the heated zone width.
            for (int i = 1; i < nx - 1; i++)
                for (int j = 1; j < ny - 1; j++)
                    tNew[i, j, 0] += rx * (tField[i + 1, j, 0] + tField[i - 1, j, 0] - 2 * tField[i, j, 0])
                                   + ry * (tField[i, j + 1, 0] + tField[i, j - 1, 0] - 2 * tField[i, j, 0]);

            // Bottom boundary
            if (_bottomBc == "adiabatic")
            {
                for (int i = 0; i < nx; i++)
                    for (int j = 0; j < ny; j++)
                        tNew[i, j, nz - 1] = tNew[i, j, nz - 2];
            }
            else
            {
                for (int i = 0; i < nx; i++)
                    for (int j = 0; j < ny; j++)
                        tNew[i, j, nz - 1] = tAmbient;
            }

            // Lateral boundaries: Adiabatic (mirror)
            for (int j = 0; j < ny; j++)
                for (int k = 0; k < nz; k++)
                {
                    tNew[0, j, k] = tNew[1, j, k];
                    tNew[nx - 1, j, k] = tNew[nx - 2, j, k];
                }
            for (int i = 0; i < nx; i++)
                for (int k = 0; k < nz; k++)
                {
                    tNew[i, 0, k] = tNew[i, 1, k];
                    tNew[i, ny - 1, k] = tNew[i, ny - 2, k];
                }

            return tNew;
        }

        /// <summary>Gaussian beam propagation: w(z) = w0 · √(1 + (z/z_R)²). DRY helper.</summary>
        private static double ComputeSpotRadiusAtZ(double z, LaserBeam laser)
        {
            double zR = laser.RayleighRange;
            return laser.SpotRadius * Math.Sqrt(1 + (z / zR) * (z / zR));
        }

        /// <summary>DoS prevention: reject grids that would exhaust memory.</summary>
        private static void CheckGridSize(int totalCells)
        {
            if (totalCells > MaxGridCells)
                throw new GridTooLargeException(totalCells, MaxGridCells);
        }

        /// <summary>Compute the smallest time step from the trajectory.</summary>
        private static double ComputeTrajectoryDt(Trajectory trajectory)
        {
            double minDt = double.MaxValue;
            for (int i = 0; i < trajectory.NPoints - 1; i++)
            {
                double segDt = trajectory.Time[i + 1] - trajectory.Time[i];
                if (segDt > 0 && segDt < minDt)
                    minDt = segDt;
            }
            return minDt < double.MaxValue ? minDt : 1e-4;
        }

        /// <summary>Find maximum depth where temperature exceeds threshold.</summary>
        private static double? FindThresholdDepth(
            double[,,] tField, int nx, int ny, int nz,
            double[] zCoords, double threshold)
        {
            int deepestK = -1;
            for (int k = 0; k < nz; k++)
            {
                double maxAtZ = double.MinValue;
                for (int i = 0; i < nx; i++)
                    for (int j = 0; j < ny; j++)
                        if (tField[i, j, k] > maxAtZ)
                            maxAtZ = tField[i, j, k];

                if (maxAtZ > threshold)
                    deepestK = k;
            }
            return deepestK >= 0 ? (double?)zCoords[deepestK] : null;
        }

        /// <summary>Generate evenly spaced array (like numpy.linspace).</summary>
        private static double[] Linspace(double start, double end, int count)
        {
            var result = new double[count];
            if (count == 1) { result[0] = start; return result; }
            double step = (end - start) / (count - 1);
            for (int i = 0; i < count; i++)
                result[i] = start + i * step;
            return result;
        }
    }
}
